### PR TITLE
attrsets: clarify that mapAttrs maps over *leaf* attrs

### PR DESCRIPTION
**Pull request by ghost**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: lib/attrsets.nix [hunk 1]
```diff
@@ -480,8 +480,10 @@ rec {
 
 
   /* Like `mapAttrs`, except that it recursively applies itself to
-     attribute sets.  Also, the first argument of the argument
-     function is a *list* of the names of the containing attributes.
+     the *leaf* attributes of a potentially-nested attribute set:
+     the second argument of the function will never be an attrset.
+     Also, the first argument of the argument function is a *list*
+     of the names of the containing attributes.
```
Comment (roberth) [inline lib/attrsets.nix:486]:
```suggestion
     of the attribute names that form the path to the leaf attribute.

     For a function that gives you control over what counts as a leaf,
     see `mapAttrsRecursiveCond`.
```

Comment (amjoseph-nixpkgs) [reply lib/attrsets.nix]: Thank you for noticing that!

**Comment by amjoseph-nixpkgs** [comment]
@roberth, related to [your suggestion](https://github.com/NixOS/nixpkgs/pull/212189#discussion_r1085210852), it appears that we do not have any kind of `mapAttrs` that applies the function to every node (i.e. no leaf/interior distinction).  Is that correct?  I went looking for one when I wrote #212186 but didn't find anything.

Is there a reason it was omitted?  Everybody who needs it seems to write their own recursion.

**Comment by roberth** [comment]
I think that's correct. It may have to do with the lack of static or nominal types, which means that we can't consistently distinguish between an attrset that's part of a tree structure and an attrset that represents something else. We do have those predicate functions (`Attrs -> Bool`, `List String -> Attrs -> Bool`) for that, but they add a lot of noise. At that point you might as well write the algorithm yourself, with a more easily understandable end result without the indirection.

> Everybody who needs it seems to write their own recursion.

In a language like Haskell, recursion schemes are kind of nice, but Nix is a different language, where different things are useful, and different things are idiomatic.

You could write a library that represents trees in a safe way, supporting a nice function like `foldTree :: (AttrSet a -> a) -> Tree -> a` which can hold any type `a` safely without a predicate function, but I think users wouldn't like the `Tree` representation.

**Review by roberth — APPROVED** [review]

**Pull request merged**